File: shared/database/src/adapters/supabase.ts
/**
 * Supabase database adapter
 * Used for test and production environments with Row-Level Security
 */

import { createClient, SupabaseClient } from '@supabase/supabase-js';
import { DatabaseAdapter, QueryResult, Transaction } from './base';

/**
 * Supabase adapter configuration
 */
export interface SupabaseAdapterConfig {
  url: string;
  anonKey: string;
  serviceRoleKey?: string;
}

/**
 * Supabase adapter implementation
 * Note: This adapter uses Supabase's PostgreSQL REST API, not raw SQL
 * For complex queries, you may need to create database functions
 */
export class SupabaseAdapter implements DatabaseAdapter {
  private client: SupabaseClient;

  /**
   * Create a new Supabase adapter
   * @param config Supabase configuration
   */
  constructor(config: SupabaseAdapterConfig) {
    // Use service role key if available (bypasses RLS for backend operations)
    // Otherwise use anon key (RLS enforced)
    const key = config.serviceRoleKey ?? config.anonKey;

    this.client = createClient(config.url, key, {
      auth: {
        persistSession: false, // Stateless backend
        autoRefreshToken: false,
      },
    });
  }

  /**
   * Execute a SQL query via Supabase RPC
   * Note: This requires a database function 'exec_sql' to be created
   * For production use, prefer using Supabase's query builder or specific RPC functions
   */
  async query<T>(sql: string, params?: any[]): Promise<QueryResult<T>> {
    // Supabase doesn't support raw SQL queries directly
    // We use RPC to execute SQL via a stored function
    const { data, error } = await this.client.rpc('exec_sql', {
      query: sql,
      params: params ?? [],
    });

    if (error) {
      throw new Error(`Supabase query error: ${error.message}`);
    }

    const rows = (data as T[]) ?? [];
    return {
      rows,
      rowCount: rows.length,
    };
  }

  /**
   * Begin a transaction
   * Note: Supabase does not support client-side transactions
   * Transactions must be handled via Edge Functions or database functions
   */
  async beginTransaction(): Promise<Transaction> {
    throw new Error(
      'Transactions not supported in Supabase adapter. '
      + 'Use database functions or Edge Functions for transactional operations.',
    );
  }

  /**
   * Close the Supabase client
   * Note: Supabase client doesn't require explicit closing
   */
  async close(): Promise<void> {
    // Supabase client handles connection management automatically
    // No explicit close needed
  }

  /**
   * Get adapter type
   */
  getType(): 'supabase' {
    return 'supabase';
  }

  /**
   * Get the underlying Supabase client
   * Use this for RLS-aware queries and Supabase-specific features
   * @returns Supabase client
   */
  getClient(): SupabaseClient {
    return this.client;
  }

  /**
   * Set the auth context (user) for RLS queries
   * @param userId User ID to set in auth context
   */
  async setAuthContext(userId: string): Promise<void> {
    // This would typically be handled by setting the JWT token
    // For service role operations, you can use SET LOCAL to set session variables
    await this.query('SET LOCAL request.jwt.claim.sub = $1', [userId]);
  }
}

File: backend/src/database/adapters/supabase.py
"""
Supabase Database Adapter

Supabase PostgreSQL implementation for test and production modes.
"""

from typing import Any
from sqlalchemy.ext.asyncio import (
    create_async_engine,
    AsyncSession,
    async_sessionmaker,
    AsyncEngine,
)
from sqlalchemy import text
from database.types import DatabaseAdapter


class SupabaseAdapter(DatabaseAdapter):
    """
    Supabase Database Adapter

    Connects to Supabase PostgreSQL database for test/production.
    Uses service role key to bypass Row Level Security (RLS).
    """

    def __init__(
        self,
        database_url: str,
        service_role_key: str,
        pool_size: int = 20,
        pool_timeout: int = 30,
    ):
        """
        Initialize Supabase adapter

        Args:
            database_url: Supabase connection string
            service_role_key: Supabase service role key (bypasses RLS)
            pool_size: Maximum connection pool size
            pool_timeout: Connection timeout in seconds
        """
        self.database_url = database_url
        self.service_role_key = service_role_key

        # Create async engine with Supabase connection
        self.engine: AsyncEngine = create_async_engine(
            database_url,
            pool_size=pool_size,
            max_overflow=10,
            pool_timeout=pool_timeout,
            echo=False,
        )

        self.session_maker = async_sessionmaker(
            self.engine, class_=AsyncSession, expire_on_commit=False
        )

    async def get_session(self) -> AsyncSession:
        """
        Get database session

        Returns:
            AsyncSession instance
        """
        return self.session_maker()

    async def close(self) -> None:
        """Close database connection pool"""
        await self.engine.dispose()

    async def execute_raw(self, query: str, params: dict[str, Any] | None = None) -> Any:
        """
        Execute raw SQL query

        Args:
            query: SQL query string
            params: Query parameters

        Returns:
            Query result
        """
        async with self.session_maker() as session:
            result = await session.execute(text(query), params or {})
            await session.commit()
            return result
